Add more documentation for assessment configuration

mwest1066 opened the pull request:
See https://prairielearn--1672.org.readthedocs.build/en/1672/ for a preview

coatless commented on doc/assessment.md:
```diff
@@ -66,9 +95,9 @@ An assessment is broken down in to a list of zones, like this:
             {
                 "numberChoose": 1,
-                "points": 5,
+                "points": [5, 3, 1],
```
I think this example should switch showing the zone implementation instead of the alternative question list, which is covered below. Maybe update it to:

```json
"zones": [
    {
        "title": "True and False",
        "questions": [
            {"id": "TFMathSymbols", "points": [1]},
            {"id": "TFQuadratic", "points": [3]},
            {"id": "TFGit", "points": [1, 0.5]}
        ]
    },
{			
        "title": "Multiple Choice",
        "questions": [
            {"id": "MCDataTypes", "points": [2, 1]},
            {"id": "MCVariableNames", "points": [2, 1]},
            {"id": "MCAtomicData", "points": [2, 1]},
            {"id": "MCExecuteCode", "points": [2, 2, 2, 1, 1, 0.5]}
        ]
    }
]
```

mwest1066 replied:
Good thinking, done!

coatless commented on doc/assessment.md:
````diff
@@ -19,36 +19,65 @@ Each assessment is a single directory in the `assessments` folder. The directory
 ```
 
-* [Format specification for assessment `infoAssessment.json`](https://github.com/PrairieLearn/PrairieLearn/blob/master/schemas/schemas/infoAssessment.json)
+[Format specification for assessment `infoAssessment.json`](https://github.com/PrairieLearn/PrairieLearn/blob/master/schemas/schemas/infoAssessment.json)
````
```suggestion
For more details, please see the [format specification for assessment `infoAssessment.json`](https://github.com/PrairieLearn/PrairieLearn/blob/master/schemas/schemas/infoAssessment.json).
```

Perhaps this should also go after the table of property values given below?

mwest1066 replied:
done

coatless commented on doc/assessment.md:
```diff
@@ -77,35 +106,133 @@ An assessment is broken down in to a list of zones, like this:
+`numberChoose` | integer | Number of questions to select for each student from this zone. (Optional; default: select all)
+`maxPoints` | number | Limit on the number of points that can be earned from this zone. (Optional; default: sum of question max points)
+`bestQuestions` | integer | Only this many questions in the zone will count towards the total points (highest-point questions will count). (Optional; default: use all questions)
+
```
Add link again to the schema specification?

```
[Format specification for zones can be found in `infoAssessment.json`](https://github.com/PrairieLearn/PrairieLearn/blob/master/schemas/schemas/infoAssessment.json)
```

mwest1066 replied:
done

coatless commented on doc/assessment.md:
```diff
@@ -77,35 +106,133 @@ An assessment is broken down in to a list of zones, like this:
+## Question alternative lists
+
+
+
```
Not sure why this code block is here? I think this is a `TODO:`?

mwest1066 replied:
yep, this goes with the `FIXME` below it 😄

coatless commented on doc/assessment.md:
```diff
@@ -77,35 +106,133 @@ An assessment is broken down in to a list of zones, like this:
+## Assessment instances, question selection, and question order
 
-* If a zone has `maxPoints`, then, of the points that are awarded for answering questions in this zone, at most `maxPoints` will count toward the total points.
+PrairieLearn distinguishes between *assessments* and *assessment instances*. An *assessment* is determined by the code in an `assessments` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the *assessment instance* for the student. There is only one copy of each assessment ...
```
How do you feel about adding a *slash* to emphasize directories? e.g. `assessments/`

mwest1066 replied:
good idea, done

coatless commented on doc/assessment.md:
```diff
@@ -77,35 +106,133 @@ An assessment is broken down in to a list of zones, like this:
+3. Concatenate the *zone questions* from each zone to form the total set of *assessment questions* for the student. This set of questions forms the *assessment instance* for this student.
 
-**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks currently in progress. Updates to `maxPoints` will take effect the next time a student grades a question. A student's “points” and “percentage score” will never decrease.
+Each zone appears in the given order in the assessment (this is not randomized for each student). Within each zone the question order is randomized per-student for Exams, but not randomized for Homeworks (but see the `shuffleQuestions` option).
```
Add link to later in the documentation for `shuffleQuestions`?

mwest1066 replied:
done

coatless commented on doc/assessment.md:
```diff
@@ -77,35 +106,133 @@ An assessment is broken down in to a list of zones, like this:
+
+## `text`: Adding text and links to assessments
+
+See the [`clientFiles` and `serverFiles`](clientServerFiles.md) page for details, and [`exam1` in the example course](https://github.com/PrairieLearn/PrairieLearn/blob/master/exampleCourse/courseInstances/Sp15/assessments/exam1/) for an example.
```
Reference FAQ entry on this topic? Or move the FAQ entry here?

coatless commented on doc/assessment.md:
```diff
@@ -77,35 +106,133 @@ An assessment is broken down in to a list of zones, like this:
+
+## Computing the total points for an assessment
+
+An assessment consists of several zones, each of which contains several questions. To compute the total assessment points, we proceed in three steps:
```
You may also want to mention here that "extra" credit can be awarded based on access rules?

coatless commented on doc/assessment.md:
```diff
@@ -117,12 +244,18 @@ Course staff see any reported issues show up on the "Issues" tab.
+## Changing assessments while students are working on them
 
-## Adding text and links to assessments
+While it is simplest to not change an assessment while students are working on it, we know that there are sometimes good reason that changes are needed. For example, a question might need to be removed because it contains an error, new questions might be added, or the points might be changed.
```
good reason -> good reasons.

mwest1066 replied:
done

coatless commented on doc/assessment.md:
```diff
@@ -117,12 +244,18 @@ Course staff see any reported issues show up on the "Issues" tab.
+While it is simplest to not change an assessment while students are working on it, we know that there are sometimes good reason that changes are needed. For example, a question might need to be removed because it contains an error, new questions might be added, or the points might be changed.
 
-See the [`clientFiles` and `serverFiles`](clientServerFiles.md) page for details, and [`exam1` in the example course](https://github.com/PrairieLearn/PrairieLearn/blob/master/exampleCourse/courseInstances/Sp15/assessments/exam1/) for an example.
+PrairieLearn’s general philosophy is to try and do the right thing in these situations. It will never automatically reduce a student’s total assessment “points” or “percentage score” due to assessment changes. Points that have been given to a student will not be taken away (unless the instructor explicitly edits the student points).
```
Smart quotes are used for `“percentage score”`. Probably want to switch to: `"percentage score"`

mwest1066 replied:
good catch, my editor was getting too clever

coatless commented on doc/assessment.md:
```diff
@@ -117,12 +244,18 @@ Course staff see any reported issues show up on the "Issues" tab.
+
+**`Exam` assessment updates:** Exam assessment instances are generated when the student starts the exam, and they are never automatically deleted, regenerated, or updated, even when the original assessment is changed in some way. This is a safety mechanism to avoid having students' assessments changed during an exam. However, if you want to force the regeneration of assessment instances then you  ...
+
+**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks that are currently in progress, and removed questions will be removed from students. Updates to `maxPoints` will take effect the next time a student grades a question.
```
Consider adding a disclaimer to speak with PrairieLearn developers and/or support staff when doing this?

mwest1066 replied:
I suspect everyone is just going to do it anyway 😄

I’ll see if I can add some more scary language warning people to not make major changes to live exams.

nwalters512 commented on doc/assessment.md:
```diff
@@ -19,93 +19,219 @@ Each assessment is a single directory in the `assessments` folder. The directory
-        ]
+            {"id": "hardQV1", "points": [5]},
+            {"id": "reallyHardQ", "points": [5, 5, 5]},
+            {“id”: “hardestQV3”, “points”: [3, 2, 1, 0.5, 0.2]}
```
```suggestion
            {"id": "hardestQV3", "points": [3, 2, 1, 0.5, 0.2]}
```

I see more smart quotes elsewhere in this document; should probably do a find/replace on the whole thing.

mwest1066 replied:
Thanks, I think I got them all now. “Helpful” editors!

nwalters512 commented:
This is awesome! Should be a massive help to instructors to have everything laid out in English in one place and easily `⌘F`-able.

nwalters512 commented on doc/assessment.md:
```diff
@@ -49,11 +50,126 @@ Type        | Description
+                    "points": [10, 9, 8, 7, 6],
+                    "alternatives": [
+                        {"id": "addVectors1"},
+                        {"id": "addVectors2"}
```
malformed JSON - missing `]`

mwest1066 replied:
fixed (plus other JSON errors)

coatless commented on mkdocs.yml:
```diff
@@ -14,6 +14,7 @@ pages:
   - 'Course configuration': 'course.md'
   - 'Course instance configuration': 'courseInstance.md'
   - 'Assessment configuration': 'assessment.md'
+  - 'Assessment design': 'assessment-design.md'
```
How about:

- `Assessment Configuration` -> `Overview of Assessment Options`
- `Assessment Design` -> `Designing an Assessment`

coatless commented on doc/assessment-design.md:
```diff
@@ -0,0 +1,126 @@
+
+## Example Homework assessment
+
+Below is an example of "Homework 1". The access rules make it always visible by TAs, and it is visible to students for full credit for 11 days (the 18th to 28th inclusive), and then is available to view and practice but for zero credit thereafter. This homework has six questions, each of which can be repeated until `maxPoints` is reached. Because it is `"type": "Homework"`, all students will see  ...
```
```suggestion
Below is an example of "Homework 1". The access rules make it always visible by TAs, and it is visible to students for full credit for 11 days (the 18th to 28th inclusive), and then is available to view and practice but for zero credit thereafter. This homework has six questions, each of which must be repeated until `maxPoints` is reached. Because it is `"type": "Homework"`, all students will see all six questions, and they can attempt them without limit. All questions are worth the same number of points because there is no advantage in using a complex point distribution on Homeworks. Students have access to a formula sheet which is the same as the one they will see on the exam, allowing them to get used to it.
```

coatless replied:
Switch **can be** to **must** for emphasize of obtaining points.

coatless commented on doc/assessment-design.md:
```diff
@@ -0,0 +1,126 @@
+
+## Example Exam assessment
+
+Below is an example of "Quiz 1" on the same topic as "Homework 1" above. The access rules always allow TAs access, and allow students access for full credit in the CBTF with a link to the given `examUuid`.
```
Mention that the `examUuid` is obtained by going to the CBTF scheduler. 

https://cbtf.engr.illinois.edu/sched

coatless commented on doc/assessment-design.md:
```diff
@@ -0,0 +1,126 @@
+            ]
+        }
+    },
+    "text": "For this quiz you can use the <a target=\"_blank\" href=\"<%= clientFilesCourse %>/formulas.pdf\">formula sheet</a>."
```
Consider adding a note on escape character use in the text to explain why `\"` is present.

coatless commented on doc/assessment-design.md:
```diff
@@ -0,0 +1,126 @@
+# Assessment design
+
```
Consider re-emphasizing two types: `Homework` and `Exam`. Otherwise, the later quiz portion might confuse folks since `type: 'Exam'` with `set: 'Quiz'`.

coatless commented:
Probably should add the descriptions from the different "retries" given earlier on Slack:


> 1. On every quiz, students receive immediate feedback about whether their answers are correct, and if they are incorrect then they can attempt that exact question again for (reduced) points. This is the setting in infoAssessment.json like "points": [10, 9, 7, 4, 2] (five attempts, 10 points if it’s correct on the first attempt, 9 on the second, etc). Essentially all courses use this “retry questions immediately” feature in PrairieLearn, because it helps students a lot if they make a small error, and it’s very easy to allow this.
> 2. Some courses choose to let students take second-chance quizzes/exams (“second chance”, “retry”, or “retake” quizzes). These are completely separate quizzes, scheduled as a second different quiz, typically taking place in a separate block of days about 1 week after the first-chance quiz. There are a variety of rules in use for how to combine the first-chance and second-chance scores, e.g., max(first-chance, 0.2*first-chance + 0.8*second-chance). These second-chance quizzes typically use different questions to the first-chance quizzes, but on the same topic. Because they need more questions, only some courses use second-chance quizzes.
>

nwalters512 commented:
@mwest1066 this has fallen pretty out-of-date. Do you intend to get this into a mergeable state?

nwalters512 commented:
@mwest1066 bump again - is this worth investing time to get it mergeable?

mwest1066 commented:
I will try and pull out the bits of this that might still be relevant. I'm not planning to try and get it mergeable as an entire PR.

reteps commented:
@mwest1066 Two comments:

We now have a schema reference for all JSON Schemas - https://prairielearn.readthedocs.io/en/latest/schemas/infoAssessment/

The JSON Schemas are now generated directly from the Zod schemas, e.g. https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/schemas/infoAssessment.ts and then you can run `make update-jsonschema` to update them.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `allowAccess`           | array   | List of access rules. (Optional; default: no student access)                                                                                                                   |
+| `zones`                 | array   | Specification of zones and questions. (Optional; default: none)                                                                                                                |
+| `text`                  | string  | HTML text shown on the assessment overview page. (Optional; default: none)                                                                                                     |
+| `multipleInstance`      | boolean | Whether to allow students to create whole new attempts at the entire assessment.(Optional; default: `false`)                                                                   |
```
```suggestion
| `multipleInstance`      | boolean | Whether to allow students to create whole new attempts at the entire assessment. (Optional; default: `false`)                                                                   |
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `text`                  | string  | HTML text shown on the assessment overview page. (Optional; default: none)                                                                                                     |
+| `multipleInstance`      | boolean | Whether to allow students to create whole new attempts at the entire assessment.(Optional; default: `false`)                                                                   |
+| `maxPoints`             | number  | The maximum points that can be earned. (Optional; default: sum of zone max points)                                                                                             |
+| `maxBonusPoints`        | number  | The maximum number of additional points that can be earned beyond `maxPoints` (Optional; default: 0)                                                                           |
```
```suggestion
| `maxBonusPoints`        | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                                                           |
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
-## Changing question-order randomization
+## Assessment naming
+
+Each assessment has a `title` describing its topic. Additionally, assessments are organized into `sets` (e.g., `Homework`, `Quiz`, `Exam`) and within each set the assessment has a `number`. Depending on the context, assessments are referred to by either an _abbreviation_, a _short name_ or a _long name_. The format of these is:
```
So that things in `code formatting` correspond to the property in JSON, which IMO is nice:

```suggestion
Each assessment has a `title` describing its topic. Additionally, each assessment belongs to a `set` (e.g., `Homework`, `Quiz`, `Exam`) and within each set the assessment has a `number`. Depending on the context, assessments are referred to by either an _abbreviation_, a _short name_ or a _long name_. The format of these is:
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
+- Long name = `[Set] [Number]: [Title]` (e.g., `Quiz 2: Coordinates and Vectors`).
 
-On `Exam` assessments, questions are randomized by default, but this can be disabled by setting `"shuffleQuestions": false` in the `infoAssessment.json` file.
+You can select a set name from the list of [standardized assessment sets](../course/index.md#standardized-assessment-sets) or create your [own](../course/index.md#adding-your-own-assessment-sets).
```
Nit, so that the link has a little more text+context for screen readers:

```suggestion
You can select a set name from the list of [standardized assessment sets](../course/index.md#standardized-assessment-sets) or [create your own](../course/index.md#adding-your-own-assessment-sets).
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
+
+## Grouping assessments by modules
+
+Instructors may want to group their assessments by course modules (topics, sections or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessme ...
```
Nit, I think we generally try to use the Oxford comma?

```suggestion
Instructors may want to group their assessments by course modules (topics, sections, or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessment will have `"module" : "Default"`.
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
+
+Instructors may want to group their assessments by course modules (topics, sections or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessme ...
+
+For more information about how to create your own assessment modules, see [Course configuration](../course/index.md#assessment-modules).
```
Nit:

```suggestion
For more information about how to create your own assessment modules, see the [course configuration](../course/index.md#assessment-modules) documentation.
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
+| `Homework`      | `false`                    | Homeworks default to using the same question order as in `infoAssessment.json` for all students.                                                                                                                               |
+| `Exam`          | `true`                     | Exams default to randomizing the question order individually for each student. This randomization happens within each zone, but the zones themselves always appear in the same order and there is no cross-zone randomization. |
+
+- If a `Homework` is set to shuffle the question order, it will use a unique-per-course number for each question, so that all students will still get the same question numbers (like #427), but they will not be in order. This makes it easy for students to discuss questions with course staff; they can say “I don't know how to do #427” and everyone will be seeing the same question #427. The main adv ...
```
Nit, no need for this to be a list.

```suggestion
If a `Homework` is set to shuffle the question order, it will use a unique-per-course number for each question, so that all students will still get the same question numbers (like #427), but they will not be in order. This makes it easy for students to discuss questions with course staff; they can say “I don't know how to do #427” and everyone will be seeing the same question #427. The main advantage of randomizing question order on Homeworks is to enable data collection on question difficulty and student behavior that is independent of the order in which questions are listed on the assessment.
```

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -105,16 +153,56 @@ An assessment is broken down in to a list of zones, like this:
 
 - If a zone has `bestQuestions`, then, of the questions in this zone, only `bestQuestions` with the highest number of awarded points will count toward the total points.
 
+## Slots for questions and question alternatives
+
+Each zone has a list of _slots_ given by the `questions` array. Each slot contains either a single question `id`:
```
We're introducing a new term here. Maybe this is the right call given that "alternative groups" or "zone questions" (which we use internally) aren't super friendly names; I just want to make sure we're thinking this through.

mwest1066 replied:
The "slot" terminology is one that I often use when giving talks about PL. But I think here it isn't really adding much, and I don't like using terminology in the docs which isn't reflected in the code. I removed "slots" in cdd94bc86d95f807483661a098282fe48aace1fc in favor of talking about "entries in the questions list".

nwalters512 commented on docs/assessment/index.md:
````diff
@@ -105,16 +153,56 @@ An assessment is broken down in to a list of zones, like this:
+}
+```
+
+- An assessment question can be specified by either a single `id` or by a list of alternatives, in which case one or more of these alternatives is chosen at random. Once the question `id` is determined, then a random variant of that question is selected. Question alternatives inherit the points of their parent group, if specified.
````
This is the second single-item list I'm seeing; is that intentional for some reason?

mwest1066 replied:
Nope, just a result of updating the PR and splitting up a list into multiple parts or deleting all but one item for some reason. Fixed in de53592c59b594e38c8e036f35a362a26d176f9f

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -250,9 +338,23 @@ As an example, suppose a student makes 4 submissions of varying degrees of corre
-## Assessment and question instances and resetting assessments
+## Assessment instances, question selection, and question order
+
+PrairieLearn distinguishes between _assessments_ and _assessment instances_. An _assessment_ is determined by the code in an `assessments/` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the _assessment instance_ for the student. There is only one copy of each assessmen ...
```
Nit, not that this really matters, but JSON is really more config.

```suggestion
PrairieLearn distinguishes between _assessments_ and _assessment instances_. An _assessment_ is determined by the configuration in an `assessments/` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the _assessment instance_ for the student. There is only one copy of each assessment, but every student has their own assessment instance.
```

nwalters512 commented on docs/uuid.md:
````diff
@@ -41,7 +35,7 @@ This script will add a UUID to the start of every JSON file that doesn't current
 ```sh
-cd <path_to_PrairieLearn>
+cd <path_to_PrairieLearn_source_code>
 cd tools
````
```suggestion
cd contrib
```

That said... what's the utility of this script? Do you think people actually use it?

reteps replied:
Can we put this information on a README in `contrib`, and splice that readme into the docs, and link to the spliced README from the uuid page?

mwest1066 replied:
I just removed it entirely. I don't think anyone is using this script 😅

nwalters512 reviewed (commented):
I haven't looked at the full rendered docs yet; here's an initial round of comments.

mwest1066 commented:
> We now have a schema reference for all JSON Schemas - https://prairielearn.readthedocs.io/en/latest/schemas/infoAssessment/ I was hoping this would become the authoritative reference for properties on schemas. I would be happy to play around with this formatting if you think that it could be improved (e.g. I noticed you used a table in this PR). Otherwise, it may make sense to update the wording on the Zod schemas directly and use those pages.

Nice! I replaced all the handwritten tables with links to schemas. The auto-generated schema docs aren't always quite as nice, but having them be correct is a huge win.

mwest1066 commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `canView`               | array   | A list of group role names that can view questions in this assessment. Only applicable for group assessments. (Optional; default: none)                                        |
+| `studentGroupCreate`    | boolean | Whether students can create groups. (Optional; default: `false`)                                                                                                               |
+| `studentGroupJoin`      | boolean | Whether students can join groups. (Optional; default: `false`)                                                                                                                 |
+| `studentGroupLeave`     | boolean | Whether students can leave groups. (Optional; default: `false`)                                                                                                                |
```
Done.

mwest1066 commented:
> Nice! I replaced all the handwritten tables with links to schemas. The auto-generated schema docs aren't always quite as nice, but having them be correct is a huge win.

And now reverted again after discussion in https://github.com/PrairieLearn/PrairieLearn/pull/11830

mwest1066 commented on docs/assessment/index.md:
```diff
@@ -105,16 +153,56 @@ An assessment is broken down in to a list of zones, like this:
+| `maxPoints`     | number  | Limit on the number of points that can be earned from this zone. (Optional; default: sum of question max points)                                      |
+| `bestQuestions` | integer | Only this many questions in the zone will count towards the total points (highest-point questions will count). (Optional; default: use all questions) |
+| `comment`       | string  | Free‑form comment for the zone (Optional; default: none)                                                                                              |
+Zone specification details are in the [format specification for `infoAssessment.json`](../schemas/infoAssessment.md)
```
Done.

reteps commented on docs/assessment/index.md:
Should the `uuid` field be linked to our UUID page?

reteps commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| Property                | Type    | Description                                                                                                                                                                    |
+| ----------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| `uuid`                  | string  | Unique identifier (UUID). E.g., `"8b4891d6-64d1-4e89-b72d-ad2133f25b2f"`. These can be obtained from [uuidgenerator.net](https://www.uuidgenerator.net) (Required; no default) |
+| `type`                  | string  | Either `"Homework"` or `"Exam"`. ([Details;](#assessment-types) Required; no default)                                                                                          |
```
Same nit in some other items

nwalters512 replied:
Changed all of these in bbb2e70. There's no longer details links at all.

reteps commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `text`                  | string  | HTML text shown on the assessment overview page. (Optional; default: none)                                                                                                     |
+| `multipleInstance`      | boolean | Whether to allow students to create whole new attempts at the entire assessment. (Optional; default: `false`)                                                                  |
+| `maxPoints`             | number  | The maximum points that can be earned. (Optional; default: sum of zone max points)                                                                                             |
+| `maxBonusPoints`        | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                                                          |
```
Do we have text to link to that talks about this?

nwalters512 replied:
Done in bbb2e70.

reteps commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+Each assessment has a `type`, which must be either `"Homework"` or `"Exam"`.
 
-For more information about how to create your own assessment modules, see [Course configuration](../course/index.md#assessment-modules).
+Homeworks are designed for formative learning and allow students to infinitely retry different random variants of questions. A persistent student can eventually get 100%.
```
Can we pull in the mastery diagram here #11839 , or should we do that part of a separate PR?

nwalters512 replied:
Please open a separate PR if you'd like to see this.

reteps commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `type`                  | string  | Either `"Homework"` or `"Exam"`. ([Details;](#assessment-types) Required; no default)                                                                                          |
+| `title`                 | string  | The title of the assessment (e.g., `"Derivatives and integrals"`). (Required; no default)                                                                                      |
+| `set`                   | string  | Which assessment set this belongs to (e.g., `"Quiz"`, `"Practice Quiz"`). (Required; no default)                                                                               |
+| `number`                | string  | The number of the assessment within the set (e.g., `"1"`, `"2B"`). (Required; no default)                                                                                      |
```
Maybe linking to / talking about the abbreviation would be nice.

nwalters512 replied:
I added a link to that section in bbb2e70.

reteps commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
+
+## Grouping assessments by modules
+
+Instructors may want to group their assessments by course modules (topics, sections, or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessm ...
```
There is some images on the concepts page about how this looks, should we duplicate those images here?

nwalters512 replied:
Let's do this in a separate PR if you'd like to see this.

reteps replied:
#12433

reteps commented on docs/assessment/index.md:
```diff
@@ -59,11 +84,34 @@ Each assessment has a `type` set to either `Homework` or `Exam`, as listed below
+| `Homework`      | `false`                    | Homeworks default to using the same question order as in `infoAssessment.json` for all students.                                                                                                                               |
+| `Exam`          | `true`                     | Exams default to randomizing the question order individually for each student. This randomization happens within each zone, but the zones themselves always appear in the same order and there is no cross-zone randomization. |
+
+If a `Homework` is set to shuffle the question order, it will use a unique-per-course number for each question, so that all students will still get the same question numbers (like #427), but they will not be in order. This makes it easy for students to discuss questions with course staff; they can say “I don't know how to do #427” and everyone will be seeing the same question #427. The main advan ...
```
Introducing the concept of a "unique-per-course question number" first would be helpful.

nwalters512 replied:
Want to do that in another PR?

reteps replied:
#12434

reteps commented on docs/assessment/index.md:
````diff
@@ -105,16 +153,57 @@ An assessment is broken down in to a list of zones, like this:
 ```
 
-- Each zone appears in the given order in the assessment. Zone titles are optional and are displayed to the student if present.
+| Zone Property   | Type    | Description                                                                                                                                           |
````
A screenshot highlighting what a zone is would be helpful, or more text about what a zone is would be good.

nwalters512 replied:
Want to add that in another PR?

reteps replied:
#12417

reteps commented on docs/assessment/index.md:
````diff
@@ -105,16 +153,57 @@ An assessment is broken down in to a list of zones, like this:
+{ "id": "hardQV1", "points": 10 }
+```
+
+Or an entry can contain a _question alternative list_:
````
Maybe explain what a `question alternative list` is?

reteps commented on docs/uuid.md:
```diff
@@ -18,38 +20,8 @@ As your course is constructed, you may run into scenarios that require you to ch
-An error like the one above can be fixed by reverting the UUID in the `infoCourseInstance.json` file back to its old value (which you can get from the `git` history).
-
-## Bulk addition of UUIDs
+- **Moving a directory: no new UUID needed.** For example, if you move an assessment between course instances or if you move a question between courses, you can leave the UUID as is. You can also rename it in this case, but it isn't necessary.
```
A table format would be nice: one of what the action is, and a check/x if a UUID is needed.

reteps replied:
#12435

reteps reviewed (approved):
I have some improvements I would want to make, but those might be out-of-scope from the original PR. Overall, it looks good.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -250,9 +339,23 @@ As an example, suppose a student makes 4 submissions of varying degrees of corre
+
+PrairieLearn distinguishes between _assessments_ and _assessment instances_. An _assessment_ is determined by the configuration in an `assessments/` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the _assessment instance_ for the student. There is only one copy of each  ...
+
+Choosing the selection of questions for each student proceeds in three steps:
```
Changed in bbb2e70.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -105,16 +153,56 @@ An assessment is broken down in to a list of zones, like this:
 
+## Question alternatives
+
+Each zone has a list of `questions`, each of which can be a single question:
```
Done in bbb2e70.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| Property                                                          | Type    | Description                                                                                                                               |
+| ----------------------------------------------------------------- | ------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+| `uuid`                                                            | string  | [Unique identifier](../uuid.md). (Required; no default)                                                                                   |
+| `type`                                                            | string  | Either `"Homework"` or `"Exam"`. ([Details](#assessment-types); Required; no default)                                                     |
```
Nope! Changed in 00f6ebd.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -105,16 +153,57 @@ An assessment is broken down in to a list of zones, like this:
+
+| Slot property     | Type            | Description                                                                                                                                                                 |
+| ----------------- | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `points`          | number or array | The number of points for this question. Can be a number (e.g., `10`) or a declining list of points (e.g., `[10, 8, 4]`) for partial credit on Exams. (Required; no default) |
```
Definitely; do you want to open a PR to add that after this lands?

jonatanschroeder replied:
#12419.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `multipleInstance`                                                | boolean | Whether to allow students to create whole new attempts at the entire assessment. (Optional; default: `false`)                             |
+| `maxPoints`                                                       | number  | The maximum points that can be earned. (Optional; default: sum of zone max points)                                                        |
+| [`maxBonusPoints`](#assessment-points)                            | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                     |
+| `shuffleQuestions`                                                | boolean | Whether to randomize the question order (Homework only). (Optional; default: `false`)                                                     |
```
Changed in 3c95960.

jonatanschroeder commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `title`                                                           | string  | The title of the assessment (e.g., `"Derivatives and integrals"`). (Required; no default)                                                 |
+| `set`                                                             | string  | Which assessment set this belongs to (e.g., `"Quiz"`, `"Practice Quiz"`). (Required; no default)                                          |
+| [`number`](#assessment-naming)                                    | string  | The number of the assessment within the set (e.g., `"1"`, `"2B"`). (Required; no default)                                                 |
+| [`module`](#grouping-assessments-by-modules)                      | string  | The module that this assessment belongs to (e.g., `Chapter 3`). (Optional; default: none)                                                 |
```
```suggestion
| [`module`](#grouping-assessments-by-modules)                      | string  | The module that this assessment belongs to (e.g., `Chapter 3`). (Optional; default: "Default")                                            |
```

nwalters512 replied:
Changed in a25143d.

jonatanschroeder commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `multipleInstance`                                                | boolean | Whether to allow students to create whole new attempts at the entire assessment. (Optional; default: `false`)                             |
+| `maxPoints`                                                       | number  | The maximum points that can be earned. (Optional; default: sum of zone max points)                                                        |
+| [`maxBonusPoints`](#assessment-points)                            | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                     |
+| [`shuffleQuestions`](#question-order-randomization)               | boolean | Whether to randomize the question order (Homework only). (Optional; default: `false`)                                                     |
```
It is no longer Homework-only.

jonatanschroeder commented on docs/assessment/index.md:
```diff
@@ -22,35 +22,60 @@ Each assessment is a single directory in the `assessments` folder or any subfold
+| `multipleInstance`                                                | boolean | Whether to allow students to create whole new attempts at the entire assessment. (Optional; default: `false`)                             |
+| `maxPoints`                                                       | number  | The maximum points that can be earned. (Optional; default: sum of zone max points)                                                        |
+| [`maxBonusPoints`](#assessment-points)                            | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                     |
+| [`shuffleQuestions`](#question-order-randomization)               | boolean | Whether to randomize the question order. (Optional; default: `false`)                                                                     |
```
The default actually depends on the assessment type. It's "False" for homework, "True" for exam.

nwalters512 commented on docs/assessment/index.md:
```diff
@@ -105,16 +154,57 @@ An assessment is broken down in to a list of zones, like this:
+
+| Slot property     | Type            | Description                                                                                                                                                                 |
+| ----------------- | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `points`          | number or array | The number of points for this question. Can be a number (e.g., `10`) or a declining list of points (e.g., `[10, 8, 4]`) for partial credit on Exams. (Required; no default) |
```
In the interest of getting this merged, I'll let you add that in a separate PR 🙂
